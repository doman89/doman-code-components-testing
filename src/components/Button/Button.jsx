import { clsx } from "clsx";

// type ButtonProps = {
//   children: ReactNode;
//   "data-testid"?: string;
//   isDisabled?: boolean;
//   isLoading?: boolean;
//   size?: "small" | "medium";
//   variant?: "solid" | "outlined";
// };

export function Button({
  children,
  isDisabled,
  isLoading,
  size = "medium",
  variant = "solid",
  ...restProps
}) {
  return (
    <button
      className={clsx([
        "button",
        {
          "button--disabled": isDisabled,
          "button--loading": isLoading,
          "button--small": size === "small",
          "button--outlined": variant === "outlined",
        },
      ])}
      disabled={isDisabled}
      {...restProps}
    >
      {children}
    </button>
  );
}
