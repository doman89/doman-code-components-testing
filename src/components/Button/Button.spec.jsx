import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { cloneElement } from "react";

import { Button } from "./Button";

describe("<Button />", () => {
  const snapshot = snapshotComponentWithTextContent("Test");

  describe("match snapshot", () => {
    it("should pass when size=medium variant=solid isLoading=false isDiabled=false", () => {
      snapshot(
        <Button
          isDisabled={false}
          isLoading={false}
          size="medium"
          variant="solid"
        />,
      );
    });

    it("should only class button when there is any additional props passed", () => {
      snapshot(<Button />);
    });

    it("should has disabled and base class when isDisabled props is passed", () => {
      snapshot(<Button isDisabled />);
    });
  });

  describe("testing interaction", () => {
    it("should be called onClick for non disabled button", async () => {
      const handleOnClick = jest.fn();
      const user = userEvent.setup();

      render(
        <Button data-testid="test" onClick={handleOnClick}>
          Test
        </Button>,
      );

      const buttonElement = screen.getByTestId("test");

      await user.click(buttonElement);
      expect(handleOnClick).toHaveBeenCalled();
    });

    it("should not be called onClick for disabled button", async () => {
      const handleOnClick = jest.fn();
      const user = userEvent.setup();

      render(
        <Button
          data-testid="test"
          isDisabled
          onClick={handleOnClick}
        >
          Test
        </Button>,
      );

      const buttonElement = screen.getByTestId("test");

      await user.click(buttonElement);
      expect(handleOnClick).not.toHaveBeenCalled();
    });
  });
});

function snapshotComponentWithTextContent(content) {
  return function (component) {
    const { asFragment, getByTestId } = render(
      cloneElement(component, {
        children: content,
        "data-testid": "test",
      }),
    );

    const renderedComponent = getByTestId("test");

    expect(renderedComponent).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  };
}

function snapshotComponent(component) {
  const { asFragment, getByTestId } = render(
    cloneElement(component, {
      "data-testid": "test",
    }),
  );

  const renderedComponent = getByTestId("test");

  expect(renderedComponent).toBeDefined();
  expect(asFragment()).toMatchSnapshot();
}
