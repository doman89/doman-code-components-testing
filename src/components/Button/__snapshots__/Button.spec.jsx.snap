// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> match snapshot should has disabled and base class when isDisabled props is passed 1`] = `
<DocumentFragment>
  <button
    class="button button--disabled"
    data-testid="test"
    disabled=""
  >
    Test
  </button>
</DocumentFragment>
`;

exports[`<Button /> match snapshot should only class button when there is any additional props passed 1`] = `
<DocumentFragment>
  <button
    class="button"
    data-testid="test"
  >
    Test
  </button>
</DocumentFragment>
`;

exports[`<Button /> match snapshot should pass when size=medium variant=solid isLoading=false isDiabled=false 1`] = `
<DocumentFragment>
  <button
    class="button"
    data-testid="test"
  >
    Test
  </button>
</DocumentFragment>
`;
